src/Microsoft.Management.Deployment/PackageManager.cpp

```diff
@@ -302,14 +308,77 @@ namespace winrt::Microsoft::Management::Deployment::implementation
                 context->Args.AddArg(Execution::Args::Type::Override, ::AppInstaller::Utility::ConvertToUTF8(options.ReplacementInstallerArguments()));
             }
         }
-        
-        AddPackageManifestToContext(packageVersionInfo, context);
 
-        // TODO: AdditionalPackageCatalogArguments is not currently supported by the underlying implementation.
+        // If the version of the package is specified use that, otherwise use the default.
+        Microsoft::Management::Deployment::PackageVersionInfo packageVersionInfo = GetPackageVersionInfo(package, options);
+        AddPackageManifestToContext(packageVersionInfo, context.get());
+
+        // Note: AdditionalPackageCatalogArguments is not needed during install since the manifest is already known so no additional calls to the source are needed. The property is deprecated.
         return context;
     }
 
-    winrt::Windows::Foundation::IAsyncOperationWithProgress<winrt::Microsoft::Management::Deployment::InstallResult, winrt::Microsoft::Management::Deployment::InstallProgress> GetInstallOperation(bool addToQueue, winrt::Microsoft::Management::Deployment::CatalogPackage package, winrt::Microsoft::Management::Deployment::InstallOptions options)
+    std::wstring GetQueueIdForInstall(winrt::Microsoft::Management::Deployment::CatalogPackage package, winrt::Microsoft::Management::Deployment::PackageCatalogInfo catalogInfo)
+    {
+        // Package Id + Catalog Id forms a unique identifier of a package for the purposes of the queue.
+        // Packages with the same id but from different catalogs can be installed at the same time.
+        std::wstring queueId;
+        queueId += package.Id();
+        queueId += std::wstring(L".");

```

Comment:
This is not a safe way to construct a unique identifier; create an object for this purpose instead.

For instance "Foo.Bar" + "Thing" == "Foo" + "Bar.Thing" and these are all legitimate.

Reply:
Also, creating a type for it makes it very explicit what its role is, gives you a way to track it in the code, and allows it to be updated FAR more easily.